store/storage/pebbledb/db.go

```diff
@@ -0,0 +1,185 @@
+	}
+
+	bzCopy := make([]byte, len(bz))
+	copy(bzCopy, bz)
+
+	return bzCopy, closer.Close()
+}
+
+func (db *Database) Set(storeKey string, version uint64, key, value []byte) error {
+	var versionBz [VersionSize]byte
+	binary.LittleEndian.PutUint64(versionBz[:], version)
+
+	batch := db.storage.NewBatch()
+
+	if err := batch.Set([]byte(latestVersionKey), versionBz[:], nil); err != nil {
+		return fmt.Errorf("failed to write PebbleDB batch: %w", err)
+	}
+	if err := batch.Set(prependStoreKey(storeKey, version, key), value, nil); err != nil {
+		return fmt.Errorf("failed to write PebbleDB batch: %w", err)
+	}
+
+	if err := batch.Commit(defaultWriteOpts); err != nil {
+		return fmt.Errorf("failed to commit PebbleDB batch: %w", err)
+	}
+
+	return nil
+}
+
+func (db *Database) Delete(storeKey string, version uint64, key []byte) error {
+	var versionBz [VersionSize]byte
+	binary.LittleEndian.PutUint64(versionBz[:], version)
+
+	batch := db.storage.NewBatch()
+
+	if err := batch.Set([]byte(latestVersionKey), versionBz[:], nil); err != nil {
+		return fmt.Errorf("failed to write PebbleDB batch: %w", err)
+	}
+	if err := batch.Delete(prependStoreKey(storeKey, version, key), nil); err != nil {
+		return fmt.Errorf("failed to write PebbleDB batch: %w", err)
+	}
+
+	if err := batch.Commit(defaultWriteOpts); err != nil {
+		return fmt.Errorf("failed to commit PebbleDB batch: %w", err)
+	}
+
+	return nil
+}
+
+func (db *Database) NewBatch(version uint64) (store.Batch, error) {
+	return NewBatch(db.storage, version)
+}
+
+func (db *Database) NewIterator(storeKey string, version uint64, start, end []byte) (store.Iterator, error) {
+	if (start != nil && len(start) == 0) || (end != nil && len(end) == 0) {
+		return nil, store.ErrKeyEmpty
+	}
+
+	if start != nil && end != nil && bytes.Compare(start, end) > 0 {
+		return nil, store.ErrStartAfterEnd
+	}
+
+	prefix := storePrefix(storeKey, version)
+	start, end = util.IterateWithPrefix(prefix, start, end)
+
+	iter := db.storage.NewIter(&pebble.IterOptions{LowerBound: start, UpperBound: end})
+	return newPebbleDBIterator(iter, prefix, start, end, false), nil
+}
+
+func (db *Database) NewReverseIterator(storeKey string, version uint64, start, end []byte) (store.Iterator, error) {
+	if (start != nil && len(start) == 0) || (end != nil && len(end) == 0) {
+		return nil, store.ErrKeyEmpty
+	}
+
+	if start != nil && end != nil && bytes.Compare(start, end) > 0 {
+		return nil, store.ErrStartAfterEnd
+	}
+
+	prefix := storePrefix(storeKey, version)
+	start, end = util.IterateWithPrefix(prefix, start, end)
+
+	iter := db.storage.NewIter(&pebble.IterOptions{LowerBound: start, UpperBound: end})
+	return newPebbleDBIterator(iter, prefix, start, end, true), nil
+}
+
+func storePrefix(storeKey string, version uint64) []byte {
+	return []byte(fmt.Sprintf(StorePrefixTpl, version, storeKey))
+}
+
+func prependStoreKey(storeKey string, version uint64, key []byte) []byte {
+	return append(storePrefix(storeKey, version), key...)
```

Comment:
I don't think this works, we need the custom comparator to be able to seek the closest version rather than looking up the exact version.

Reply:
Yes, you're totally right and this is a doubt I had during impl. Looking at the MVCC example 👍

Reply:
@yihuang I pushed some commits to setup the Comparer. Could you take a quick look to see if it's in the right direction.

Reply:
I still need to update the CRUD methods along with the iterator FYI